src/coverlet.tracker/CoverageTracker.cs

```diff
@@ -25,7 +26,7 @@ public static void MarkExecuted(string file, string evt)
         {
             if (t_events == null)
             {
-                t_events = new Dictionary<string, Dictionary<string, int>>();
+                t_events = new Dictionary<string, Dictionary<string, int>>(_stringHashSuffixComparer);
```

Comment:
You know, now that I think of it there's really no reason to store the path to the hits file as a key in a dictionary. By design, the hits file for an assembly remains the same for an entire test session. Also, the `coverage.tracker` assembly is usually loaded by the runtime in different contexts for each assembly that depends on it so there's no chance of multiple instrumented assemblies supplying conflicting information.

I believe this approach was added when we had multiple separate gzipped hits files because some users were experiencing considerable disk usage due to large hit files. That's no longer the case.

In this case we can simply make the hits file a field and assign to it once (if it evaluates to `null`) and have a standalone `Dictionary<string, int>` to store the marker information.

Let me know what you think

Reply:
It sounds correct to me. I'll give a try during the day.

Reply:
This requires some risky changes: the same thread can receive events from different modules. In other words while the hits file is the same for a given assembly the same thread may receive markers from different assemblies. I'm going to try some minimal changes to optimize that but perhaps we should go for the safe and simple optimization in the short run, while we work on a more profound one.

Reply:
I don't think it's currently possible for the `coverlet.tracker` assembly to receive markers from different assemblies. Here's an illustration of how I believe things are setup in a multiple assembly scenario.

1. `SomeAssembly.Tests.dll` depends on `SomeAssembly1.dll` and `SomeAssembly2.dll` which are both instrument-able
2. `SomeAssembly1.dll` and `SomeAssembly2.dll` are instrumented by coverlet and the `coverlet.tracker.dll` dependency is copied to the same folder the two assemblies occupy.
3. The test project `SomeAssembly.Tests` is run and during that process, methods in `SomeAssembly1.dll` are called which contain instrumentation instructions that call out to `coverlet.tracker.dll`
4. During that same process, instrumented methods from `SomeAssembly2.dll` also call out to `coverlet.tracker.dll`

What happens in this case is that both `SomeAssembly1.dll` and `SomeAssembly2.dll` have `coverlet.tracker.dll` loaded in different contexts. Whatever changes `SomeAssembly1.dll` makes to static variables in its own copy of `coverlet.tracker.dll` doesn't interfere with how `SomeAssembly2.dll` interacts with its own equivalent copy. `SomeAssembly1.dll` and `SomeAssembly2.dll` depend independently on `coverlet.tracker.dll` the same way two different projects can depend on the same version of `Newtonsoft.Json` and still not get in each other's way even if their binaries all end up in the same folder.

See the original naive implementation of the `CoverageTracker` class https://github.com/tonerdo/coverlet/blob/1d92ee66a1bb68629bf31022a300e57c36ce535f/src/coverlet.core/CoverageTracker.cs. It worked perfectly, it just wasn't performant or thread safe

Reply:
I did a test to be sure: check https://github.com/tonerdo/coverlet/compare/master...pjanotti:test.multithread?expand=1 It does hit the exception when running the performance test. The thread static is the same in a single app domain (that even doesn't really exist on core).

Reply:
Does this happen only in a multi-threaded scenario?

Reply:
No, single thread too, you just need different modules on the same thread.

Reply:
I did an experiment to generate a "tracker type" per module and get rid of the dictionaries, it seems to be going well. It needs more testing to ensure correctness and that it doesn't break some scenarios that are working now. If you want to take a peak before I do some clean-up/polishing/fine tunning take a look [here]https://github.com/tonerdo/coverlet/compare/master...pjanotti:tracker.type.per.module?expand=1(). There is a bit of trade-off and it probably it will use more memory in projects with lower coverage numbers but for the project that I tested here it consistently runs the whole test in ~19 seconds (about 6 times faster than the optimization proposed on this PR).

Reply:
Took a quick peek. Seems like a good crop of changes. Will merge this in now since I'm fine with it, thanks for the change and explanations

Reply:
Thanks @tonerdo